tests/python/pants_test/engine/legacy/test_console_rule_integration.py

```diff
@@ -49,3 +56,46 @@ def test_v2_goal_validation_both(self):
       ':',
       success=True
     )
+
+  def test_v2_list_loop(self):
+    # Create a BUILD file in a nested temporary directory, and add additional targets to it.
+    with self.pantsd_test_context() as (workdir, config, checker), \
+        temporary_dir(root_dir=get_buildroot()) as tmpdir:
+      rel_tmpdir = fast_relpath(tmpdir, get_buildroot())
+
+      def dump(content):
+        safe_file_dump(os.path.join(tmpdir, 'BUILD'), content)
+
+      # Dump an initial target before starting the loop.
+      dump('target(name="one")')
+
+      # Launch the loop as a background process.
+      handle = self.run_pants_with_workdir_without_waiting(
+        [
+          '--no-v1',
+          '--v2',
+          '--loop',
+          '--loop-max=3',
+          'list',
+          '{}:'.format(tmpdir),
+        ],
+        workdir,
+        config,
+      )
+
+      # Wait for the loop to stabilize.
+      time.sleep(10)
+      checker.assert_started()
+
+      # Replace the BUILD file content twice.
+      dump('target(name="two")')
+      time.sleep(5)
+      dump('target(name="three")')
+
+      # Verify that the three different target states were listed, and that the process exited.
+      pants_result = handle.join()
+      self.assert_success(pants_result)
+      self.assertEquals(
+          ['{}:{}'.format(rel_tmpdir, name) for name in ('one', 'two', 'three')],
```

Comment:
I'm slightly confused by this expectation; I like it, but why isn't the expectation:
```
one
one
two
one
two
three
```
? What happened to the previous results?

Reply:
Each dump is an overwrite, so in each loop 1 target is listed. First it's one, then two, then three.

Reply:
Yep.

Reply:
Since the same BUILD file is overwritten twice with new contents in each overwrite the list task, which requests addresses from the engine will get a single new address each time. In other words the overwrites invalidate (a portion of) the build graph and it's exactly this portion which is used to produce the output of list each of the three times it is run.

Reply:
One possible sequence of events is:
1. user creates a BUILD file containing a target named `one`
2. user runs pants list, sees `one`
3. user renames the target from `one` to `two`
4. user runs pants list, sees `two` (which they're expecting, because they renamed the target)
5. etc.

That's all this is, except the loop is running `list` for you when it sees that files have changed.

Why does the user see `"two"` at step 4? Because the engine invalidates Nodes in the graph when you change files.

Reply:
Aha, for some reason I thought this was appending, not replacing!